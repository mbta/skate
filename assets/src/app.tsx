// We need to import the CSS so that webpack will load it.
// The MiniCssExtractPlugin is used to separate it out into
// its own CSS file.
declare function require(name: string): string
require("../css/app.scss")

// webpack automatically bundles all modules in your
// entry points. Those entry points can be configured
// in "webpack.config.js".
//
// Import dependencies
//
import "core-js/stable"
import "leaflet-defaulticon-compatibility/dist/leaflet-defaulticon-compatibility.webpack.css" // see https://github.com/Leaflet/Leaflet/issues/4968#issuecomment-483402699
import "phoenix_html"
import * as React from "react"
import { createRoot } from "react-dom/client"
import sentryInit from "./helpers/sentryInit"
import AppStateWrapper from "./components/appStateWrapper"
import { tagManagerIdentify } from "./helpers/googleTagManager"
import { fullStoryInit } from "./helpers/fullStory"
import inTestGroup, { TestGroups } from "./userInTestGroup"

document.documentElement.dataset.demoMode = inTestGroup(
  TestGroups.DemoMode
).toString()

const username = document
  .querySelector("meta[name=username]")
  ?.getAttribute("content")

fullStoryInit(window.fullStoryInitialization?.organizationId, username)

sentryInit(
  window.sentryInitialization?.initArgs,
  username || undefined,
  window.sentryInitialization?.orgSlug
)

const userUuid = document
  .querySelector("meta[name=user-uuid]")
  ?.getAttribute("content")

tagManagerIdentify(userUuid)

// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
const root = createRoot(document.getElementById("app")!)
root.render(<AppStateWrapper />)
