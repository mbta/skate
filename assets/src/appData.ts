const appData = (): DOMStringMap | undefined => {
  const dataEl = document.getElementById("app")

  if (!dataEl) {
    return undefined
  }

  return dataEl.dataset
}

export default appData
