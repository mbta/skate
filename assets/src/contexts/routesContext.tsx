import React, { createContext, ReactElement, useContext } from "react"
import { Route, RouteId } from "../schedule"

const RoutesContext = createContext<Route[] | null>(null)

export const useRoute = (routeId: RouteId | null | undefined): Route | null => {
  const routes: Route[] | null = useContext(RoutesContext)
  return (routes || []).find((route) => route.id === routeId) || null
}

export const RoutesProvider = ({
  routes,
  children,
}: {
  routes: Route[] | null
  children: ReactElement<HTMLElement>
}) => {
  return (
    <RoutesContext.Provider value={routes}>{children}</RoutesContext.Provider>
  )
}

export default RoutesContext
