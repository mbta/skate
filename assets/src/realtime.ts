import {
  BlockId,
  DirectionId,
  RouteId,
  RoutePatternId,
  StopId,
  TimepointId,
  TripId,
  ViaVariant,
} from "./schedule"

import { Crowding } from "./models/crowding"
import { DetourId } from "./models/detoursList"

export interface BlockWaiver {
  startTime: Date
  endTime: Date
  causeId: number
  causeDescription: string
  remark: string | null
}

export interface DataDiscrepancy {
  attribute: string
  sources: DataDiscrepancySource[]
}

export interface DataDiscrepancySource {
  id: SourceId
  value: string | null
}

export type EndOfTripType = "another_trip" | "swing_off" | "pull_back"

export interface Ghost {
  id: VehicleId
  directionId: DirectionId
  routeId: RouteId
  routePatternId: RoutePatternId | null
  tripId: TripId
  headsign: string
  blockId: BlockId
  runId: RunId | null
  viaVariant: ViaVariant | null
  incomingTripDirectionId: DirectionId | null
  layoverDepartureTime: number | null
  scheduledTimepointStatus: VehicleTimepointStatus | null
  scheduledLogonTime: number | null
  routeStatus: RouteStatus
  blockWaivers: BlockWaiver[]
  currentPieceStartPlace: string | null
  currentPieceFirstRoute: RouteId | null
}

export type NotificationId = string

// A Notification's `content` is populated with an Elixir struct and the
// `__struct__` field is serialized so that the type can be discerned on the
// frontend.
export enum NotificationType {
  BridgeMovement = "Elixir.Notifications.Db.BridgeMovement",
  BlockWaiver = "Elixir.Notifications.Db.BlockWaiver",
  Detour = "Elixir.Notifications.Db.Detour",
}

export interface BridgeLoweredNotification {
  $type: NotificationType.BridgeMovement
  status: "lowered"
}

export interface BridgeRaisedNotification {
  $type: NotificationType.BridgeMovement
  status: "raised"
  loweringTime: Date
}

export type BridgeNotification =
  | BridgeLoweredNotification
  | BridgeRaisedNotification

export type BlockWaiverNotification = {
  $type: NotificationType.BlockWaiver
  createdAt: Date
  reason: BlockWaiverReason
  routeIds: RouteId[]
  runIds: RunId[]
  tripIds: TripId[]
  operatorId: string | null
  operatorName: string | null
  routeIdAtCreation: string | null
  startTime: Date
  endTime: Date | null
}

export enum DetourNotificationStatus {
  Activated = "activated",
  Deactivated = "deactivated",
}

export type DetourNotification = {
  $type: NotificationType.Detour
  status: DetourNotificationStatus
  detourId: DetourId
  headsign: string
  route: string
  direction: string
  origin: string
}

export type NotificationContentTypes =
  | BridgeNotification
  | BlockWaiverNotification
  | DetourNotification

export interface Notification<
  TNotification extends NotificationContentTypes = NotificationContentTypes
> {
  id: NotificationId
  createdAt: Date
  state: NotificationState
  content: TNotification
}

export function isBlockWaiverNotification(
  notification: Notification
): notification is Notification<BlockWaiverNotification> {
  return notification.content.$type === NotificationType.BlockWaiver
}
export type BlockWaiverReason =
  | "other"
  | "accident"
  | "adjusted"
  | "disabled"
  | "diverted"
  | "manpower"
  | "operator_error"
  | "traffic"

export type NotificationState = "unread" | "read" | "deleted"

export type SourceId = string

export type RouteStatus = "on_route" | "laying_over" | "pulling_out"

export type RunId = string

export interface ByRunId<T> {
  [runId: string]: T
}

export type TrainVehicleId = string

export interface TrainVehicle {
  id: TrainVehicleId
  latitude: number
  longitude: number
  bearing: number
}

export interface Vehicle {
  id: VehicleId
  label: string | null
  runId: RunId | null
  timestamp: number
  latitude: number
  longitude: number
  directionId: DirectionId | null
  routeId: RouteId | null
  routePatternId: RoutePatternId | null
  tripId: TripId | null
  headsign: string | null
  viaVariant: ViaVariant | null
  operatorId: string | null
  operatorFirstName: string | null
  operatorLastName: string | null
  operatorLogonTime: Date | null
  overloadOffset?: number
  bearing?: number
  blockId: BlockId | null
  previousVehicleId: string | null
  scheduleAdherenceSecs: number | null
  incomingTripDirectionId: DirectionId | null
  isShuttle: boolean
  isOverload: boolean
  isOffCourse: boolean
  isRevenue: boolean | null
  layoverDepartureTime: number | null
  pullbackPlaceName: string | null
  dataDiscrepancies: DataDiscrepancy[]
  stopStatus: VehicleStopStatus
  timepointStatus: VehicleTimepointStatus | null
  scheduledLocation: VehicleScheduledLocation | null
  routeStatus: RouteStatus
  endOfTripType: EndOfTripType
  blockWaivers: BlockWaiver[]
  crowding: Crowding | null
}

export interface VehicleInScheduledService extends Vehicle {
  directionId: DirectionId
  routeId: RouteId
  blockId: BlockId
}

export type VehicleId = string

export interface VehicleStopStatus {
  stopId: StopId | null
  stopName: string | null
}

export interface VehicleScheduledLocation {
  routeId: RouteId
  routePatternId: RoutePatternId | null
  directionId: DirectionId
  tripId: TripId
  runId: RunId
  timeSinceTripStartTime: number
  headsign: string | null
  viaVariant: ViaVariant | null
  timepointStatus: VehicleTimepointStatus
}

export interface VehicleTimepointStatus {
  timepointId: TimepointId
  fractionUntilTimepoint: number
}
