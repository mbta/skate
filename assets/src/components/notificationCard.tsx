import React, { ReactElement } from "react"
import { useRoute, useRoutes } from "../contexts/routesContext"
import {
  BlockWaiverNotification,
  Notification,
  BlockWaiverReason,
  NotificationType,
  isBlockWaiverNotification,
} from "../realtime"
import { Route } from "../schedule"
import { formattedTime } from "../util/dateTime"
import { CardBody, CardProperties, CardReadable } from "./card"
import { fullStoryEvent } from "../helpers/fullStory"

export const NotificationCard = ({
  notification,
  currentTime,
  openVPPForCurrentVehicle,
  hideLatestNotification,
  noFocusOrHover,
}: {
  notification: Notification
  currentTime: Date
  openVPPForCurrentVehicle: (notification: Notification) => void
  hideLatestNotification?: () => void
  noFocusOrHover?: boolean
}): ReactElement<HTMLElement> => {
  const routes = useRoutes(
    isBlockWaiverNotification(notification) ? notification.content.routeIds : []
  )
  const routeAtCreation = useRoute(
    isBlockWaiverNotification(notification)
      ? notification.content.routeIdAtCreation
      : null
  )
  const isUnread = notification.state === "unread"
  return (
    <CardReadable
      currentTime={currentTime}
      title={<>{title(notification)}</>}
      style="kiwi"
      isActive={isUnread}
      openCallback={() => {
        openVPPForCurrentVehicle(notification)

        if (hideLatestNotification) {
          hideLatestNotification()
        }

        if (notification.content.$type === NotificationType.BridgeMovement) {
          fullStoryEvent("User clicked Chelsea Bridge Notification", {})
        }
      }}
      closeCallback={hideLatestNotification}
      time={notification.createdAt}
      noFocusOrHover={noFocusOrHover}
    >
      <CardBody>{description(notification, routes, routeAtCreation)}</CardBody>
      {isBlockWaiverNotification(notification) && (
        <CardProperties
          properties={[
            {
              label: "Run",
              value:
                notification.content.runIds.length > 0
                  ? notification.content.runIds.join(", ")
                  : null,
            },
            {
              label: "Operator",
              value:
                notification.content.operatorName !== null &&
                notification.content.operatorId !== null
                  ? `${notification.content.operatorName} #${notification.content.operatorId}`
                  : null,
              sensitive: true,
            },
          ]}
        />
      )}
    </CardReadable>
  )
}

export const title = (notification: Notification) => {
  switch (notification.content.$type) {
    case NotificationType.BlockWaiver: {
      return blockWaiverNotificationTitle(notification.content.reason)
    }
    case NotificationType.BridgeMovement: {
      switch (notification.content.status) {
        case "lowered":
          return "Chelsea St Bridge Lowered"
        case "raised":
          return "Chelsea St Bridge Raised"
      }
    }
  }
}
export const blockWaiverNotificationTitle = (
  reason: BlockWaiverReason
): string => {
  switch (reason) {
    case "manpower":
      return "No Operator"
    case "diverted":
      return "Diversion"
    default:
      return reason
        .replace("_", " ")
        .split(" ")
        .map((word) => word.charAt(0).toUpperCase() + word.slice(1))
        .join(" ")
  }
}
const blockWaiverDescription = (
  notification: BlockWaiverNotification,
  routes: Route[],
  routeAtCreation: Route | null
): string => {
  const routeNames = routes.map((route) => route.name).join(", ")
  const routeNameAtCreation = routeAtCreation
    ? routeAtCreation.name
    : notification.routeIdAtCreation

  switch (notification.reason) {
    case "manpower":
      return `OCC reported that an operator is not available on the ${routeNames}.`
    case "disabled":
      return `OCC reported that a vehicle is disabled on the ${
        routeNameAtCreation || routeNames
      }.`
    case "diverted":
      return `OCC reported that an operator has been diverted from the ${routeNames}.`
    case "accident":
      return `OCC reported that an operator has been in an accident on the ${
        routeNameAtCreation || routeNames
      }.`
    case "adjusted":
      return `OCC reported an adjustment on the ${routeNames}.`
    case "operator_error":
      return `OCC reported an operator error on the ${
        routeNameAtCreation || routeNames
      }.`
    case "traffic":
      return `OCC created a dispatcher note due to traffic on the ${
        routeNameAtCreation || routeNames
      }.`
    case "other":
    default:
      return `OCC created a dispatcher note for the ${routeNames}.`
  }
}

const description = (
  notification: Notification,
  routes: Route[],
  routeAtCreation: Route | null
): string => {
  switch (notification.content.$type) {
    case NotificationType.BlockWaiver: {
      return blockWaiverDescription(
        notification.content,
        routes,
        routeAtCreation
      )
    }
    case NotificationType.BridgeMovement: {
      switch (notification.content.status) {
        case "raised":
          return `OCC reported that the Chelsea St bridge will be raised until ${formattedTime(
            notification.content.loweringTime
          )}.`

        case "lowered":
          return "OCC reported that the Chelsea St bridge has been lowered."
      }
    }
  }
}
