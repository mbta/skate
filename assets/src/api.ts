import "whatwg-fetch"
import appData from "./appData"
import { Block, Run } from "./minischedule"
import { reload } from "./models/browser"
import { blockFromData, runFromData } from "./models/minischeduleData"
import { swingsFromData } from "./models/swingsData"
import { NotificationId, NotificationState } from "./realtime.d"
import { RouteSettings } from "./routeSettings"
import {
  DirectionName,
  Route,
  RouteId,
  Shape,
  Swing,
  Timepoint,
  TripId,
} from "./schedule.d"

export interface RouteData {
  id: string
  direction_names: {
    "0": DirectionName
    "1": DirectionName
  }
  name: string
}

const checkResponseStatus = (response: Response) => {
  if (response.status === 200) {
    return response
  }

  if (Math.floor(response.status / 100) === 3 || response.status === 403) {
    // If the API sends us a redirect or forbidden, the user needs to
    // re-authenticate. Reload to go through the auth flow again.
    reload(true)
  }

  throw new Error(`Response error: ${response.status}`)
}

const parseJson = (response: Response) => response.json()

export const apiCall = <T>({
  url,
  parser,
  defaultResult,
}: {
  url: string
  parser: (data: any) => T
  defaultResult?: T
}): Promise<T> =>
  fetch(url)
    .then(checkResponseStatus)
    .then(parseJson)
    .then(({ data: data }: { data: any }) => parser(data))
    .catch((error) => {
      if (defaultResult === undefined) {
        throw error
      } else {
        return defaultResult
      }
    })

export const parseRouteData = ({
  id,
  direction_names,
  name,
}: RouteData): Route => ({
  id,
  directionNames: direction_names,
  name,
})

const parseRoutesData = (routesData: RouteData[]): Route[] =>
  routesData.map(parseRouteData)

export const fetchRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/routes",
    parser: parseRoutesData,
  })

export const fetchShapeForRoute = (routeId: RouteId): Promise<Shape[]> =>
  apiCall({
    url: `/api/shapes/route/${routeId}`,
    parser: (shapes: Shape[]) => shapes,
    defaultResult: [],
  })

export const fetchShapeForTrip = (tripId: TripId): Promise<Shape | null> =>
  apiCall({
    url: `/api/shapes/trip/${tripId}`,
    parser: (shape: Shape | null) => shape,
    defaultResult: null,
  })

export const fetchShuttleRoutes = (): Promise<Route[]> =>
  apiCall({
    url: "/api/shuttles",
    parser: parseRoutesData,
    defaultResult: [],
  })

export const fetchTimepointsForRoute = (
  routeId: RouteId
): Promise<Timepoint[]> =>
  apiCall({
    url: `/api/routes/${routeId}`,
    parser: (timepoints: Timepoint[]) => timepoints,
    defaultResult: [],
  })

export const fetchMinischeduleRun = (tripId: TripId): Promise<Run | null> =>
  apiCall({
    url: `/api/minischedule/run/${tripId}`,
    parser: nullableParser(runFromData),
    defaultResult: null,
  })

export const fetchMinischeduleBlock = (tripId: TripId): Promise<Block | null> =>
  apiCall({
    url: `/api/minischedule/block/${tripId}`,
    parser: nullableParser(blockFromData),
    defaultResult: null,
  })

export const fetchNearestIntersection = (
  latitude: number,
  longitude: number
): Promise<string | null> =>
  apiCall({
    url: `/api/intersection?latitude=${latitude}&longitude=${longitude}`,
    parser: nullableParser((intersection: string) => intersection),
    defaultResult: null,
  })

export const fetchSwings = (): Promise<Swing[] | null> =>
  apiCall({
    url: "/api/swings",
    parser: nullableParser(swingsFromData),
    defaultResult: null,
  })

export const putNotificationReadState = (
  newReadState: NotificationState,
  notificationIds: NotificationId[]
): void => {
  const url = `/api/notification_read_state?new_state=${newReadState}&notification_ids=${notificationIds.join(
    ","
  )}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putUserSetting = (field: string, value: string): void => {
  const url = `/api/user_settings?field=${field}&value=${value}`
  fetch(url, {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
  })
}

export const putRouteSettings = (routeSettings: RouteSettings): void => {
  fetch("/api/route_settings", {
    method: "PUT",
    headers: {
      "Content-Type": "application/json",
      "x-csrf-token": getCsrfToken(),
    },
    body: JSON.stringify(routeSettings),
  })
}

const getCsrfToken = (): string => appData()?.csrfToken || ""

export const nullableParser =
  <Data, T>(parser: (data: Data) => T): ((data: Data | null) => T | null) =>
  (data: Data | null) =>
    data === null ? null : parser(data)
