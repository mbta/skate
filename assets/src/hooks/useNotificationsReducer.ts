import { Dispatch as ReactDispatch, useReducer } from "react"
import { putNotificationReadState } from "../api"
import { otherNotificationReadState } from "../contexts/notificationsContext"
import {
  NotificationData,
  notificationFromData,
} from "../models/notificationData"
import { Notification } from "../realtime.d"

export interface State {
  notifications: Notification[]
  showLatestNotification: boolean
}

export const initialState: State = {
  notifications: [],
  showLatestNotification: false,
}

interface AddNotificationAction {
  type: "ADD_NOTIFICATION"
  payload: { notification: Notification }
}

export const addNotification = (
  notification: Notification
): AddNotificationAction => ({
  type: "ADD_NOTIFICATION",
  payload: { notification },
})

interface ExpireNotificationsAction {
  type: "EXPIRE_NOTIFICATIONS"
  payload: { now: Date }
}

export const expireNotifications = (now: Date): ExpireNotificationsAction => ({
  type: "EXPIRE_NOTIFICATIONS",
  payload: { now },
})

interface HideLatestNotificationAction {
  type: "HIDE_LATEST_NOTIFICATION"
}

export const hideLatestNotification = (): HideLatestNotificationAction => ({
  type: "HIDE_LATEST_NOTIFICATION",
})

interface MarkAllAsReadAction {
  type: "MARK_ALL_AS_READ"
}

export const markAllAsRead = (): MarkAllAsReadAction => ({
  type: "MARK_ALL_AS_READ",
})

interface SetNotificationsAction {
  type: "SET_NOTIFICATIONS"
  payload: {
    notificationsData: NotificationData[]
    isInitialLoad: boolean
  }
}

export const setNotifications = (
  notificationsData: NotificationData[],
  isInitialLoad: boolean
): SetNotificationsAction => ({
  type: "SET_NOTIFICATIONS",
  payload: { notificationsData, isInitialLoad },
})

interface ToggleReadStateAction {
  type: "TOGGLE_READ_STATE"
  payload: { notification: Notification }
}

export const toggleReadState = (
  notification: Notification
): ToggleReadStateAction => ({
  type: "TOGGLE_READ_STATE",
  payload: { notification },
})

export type Action =
  | AddNotificationAction
  | ExpireNotificationsAction
  | HideLatestNotificationAction
  | MarkAllAsReadAction
  | SetNotificationsAction
  | ToggleReadStateAction

export type Dispatch = ReactDispatch<Action>

type Reducer = (state: State, action: Action) => State

export const notificationsReducer = (
  notifications: Notification[],
  action: Action
): Notification[] => {
  switch (action.type) {
    case "ADD_NOTIFICATION":
      const newNotification = (action as AddNotificationAction).payload
        .notification
      return [newNotification, ...notifications]
    case "EXPIRE_NOTIFICATIONS":
      return notifications.filter((notification) => {
        const maxAgeInMs = 8 * 60 * 60 * 1000
        const now = (action as ExpireNotificationsAction).payload.now
        const ageInMs = now.valueOf() - notification.createdAt.valueOf()
        return ageInMs < maxAgeInMs
      })
    case "MARK_ALL_AS_READ":
      return notifications.map((notification) => ({
        ...notification,
        state: "read",
      }))
    case "SET_NOTIFICATIONS":
      const notificationsData = (action as SetNotificationsAction).payload
        .notificationsData
      return notificationsData.map(notificationFromData)
    case "TOGGLE_READ_STATE":
      const notificationToToggle = (action as ToggleReadStateAction).payload
        .notification
      return notifications.map((notification) =>
        notification.id === notificationToToggle.id
          ? {
              ...notification,
              state: otherNotificationReadState(notification.state),
            }
          : notification
      )

    default:
      return notifications
  }
}

const showLatestNotificationReducer = (
  showLatestNotification: boolean,
  action: Action
): boolean => {
  switch (action.type) {
    case "ADD_NOTIFICATION":
      return true
    case "HIDE_LATEST_NOTIFICATION":
      return false
    case "SET_NOTIFICATIONS":
      const isInitialLoad = (action as SetNotificationsAction).payload
        .isInitialLoad
      if (isInitialLoad) {
        return true
      }
      return showLatestNotification
    default:
      return showLatestNotification
  }
}

export const reducer: Reducer = (state: State, action: Action): State => ({
  notifications: notificationsReducer(state.notifications, action),
  showLatestNotification: showLatestNotificationReducer(
    state.showLatestNotification,
    action
  ),
})

const deliverFullstoryEvent = (numStacked: number): void => {
  if (window.FS && window.username) {
    window.FS.event("Notification delivered", {
      num_stacked_int: numStacked,
    })
  }
}

const persistMarkAllAsRead = (notifications: Notification[]): void => {
  const notificationIds = notifications.map((notification) => notification.id)
  putNotificationReadState("read", notificationIds)
}

const persistToggledNotificationReadState = (
  notification: Notification
): void => {
  putNotificationReadState(otherNotificationReadState(notification.state), [
    notification.id,
  ])
}

export const useNotificationsReducer = (): [State, Dispatch] => {
  const [state, dispatch] = useReducer(reducer, initialState)

  const dispatchWithSideEffects: Dispatch = (action: Action): void => {
    switch (action.type) {
      case "ADD_NOTIFICATION":
        deliverFullstoryEvent(state.notifications.length + 1)
        break
      case "MARK_ALL_AS_READ":
        persistMarkAllAsRead(state.notifications)
        break
      case "TOGGLE_READ_STATE":
        const notificationToToggle = (action as ToggleReadStateAction).payload
          .notification
        persistToggledNotificationReadState(notificationToToggle!)
        break
    }

    dispatch(action)
  }

  return [state, dispatchWithSideEffects]
}

export default useNotificationsReducer
