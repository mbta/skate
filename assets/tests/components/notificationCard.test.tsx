import React from "react"
import { render } from "@testing-library/react"
import { NotificationCard } from "../../src/components/notificationCard"
import { Notification, NotificationReason } from "../../src/realtime"
import notificationFactory from "../factories/notification"
import routeFactory from "../factories/route"
import userEvent from "@testing-library/user-event"
import { hideLatestNotification } from "../../src/hooks/useNotificationsReducer"
import { RoutesProvider } from "../../src/contexts/routesContext"
import { mockFullStoryEvent } from "../testHelpers/mockHelpers"

const notification = notificationFactory.build({
  routeIds: ["route1", "route2"],
  routeIdAtCreation: null,
})

const notificationWithMatchedVehicle: Notification = {
  ...notification,
  operatorName: "operatorName",
  operatorId: "operatorId",
  routeIdAtCreation: "route1",
}

const routes = [
  routeFactory.build({
    id: "route1",
    name: "r1",
  }),
  routeFactory.build({
    id: "route2",
    name: "r2",
  }),
  routeFactory.build({
    id: "route3",
    name: "r3",
  }),
]

describe("NotificationCard", () => {
  test("transforms reasons into human-readable titles", () => {
    const n: Notification = { ...notification, reason: "operator_error" }
    const result = render(
      <RoutesProvider routes={routes}>
        <NotificationCard
          notification={n}
          currentTime={new Date()}
          openVPPForCurrentVehicle={jest.fn()}
        />
      </RoutesProvider>
    )
    expect(result.queryByText(/Operator Error/)).not.toBeNull()
  })

  test("uses custom titles if available", () => {
    const n: Notification = { ...notification, reason: "manpower" }
    const result = render(
      <RoutesProvider routes={routes}>
        <NotificationCard
          notification={n}
          currentTime={new Date()}
          openVPPForCurrentVehicle={jest.fn()}
        />
      </RoutesProvider>
    )
    expect(result.queryByText(/No Operator/)).not.toBeNull()
  })

  test("renders a notification with an unexpected reason", () => {
    const n: Notification = { ...notification, reason: "other" }
    const result = render(
      <RoutesProvider routes={routes}>
        <NotificationCard
          notification={n}
          currentTime={new Date()}
          openVPPForCurrentVehicle={jest.fn()}
        />
      </RoutesProvider>
    )

    expect(result.getByText(/Other/)).not.toBeNull()
  })

  test("routeIdAtCreation shows when relevant", () => {
    const n: Notification = {
      ...notificationWithMatchedVehicle,
      reason: "accident",
      routeIds: ["route2", "route3"],
      routeIdAtCreation: "route1",
    }
    const result = render(
      <RoutesProvider routes={routes}>
        <NotificationCard
          notification={n}
          currentTime={new Date()}
          openVPPForCurrentVehicle={jest.fn()}
        />
      </RoutesProvider>
    )
    expect(result.getByText(/r1/)).not.toBeNull()
  })

  test("falls back to affected routeIds if routeIdAtCreation is missing", () => {
    const n: Notification = {
      ...notification,
      reason: "accident",
      routeIds: ["route2", "route3"],
      routeIdAtCreation: null,
    }
    const result = render(
      <RoutesProvider routes={routes}>
        <NotificationCard
          notification={n}
          currentTime={new Date()}
          openVPPForCurrentVehicle={jest.fn()}
        />
      </RoutesProvider>
    )
    expect(result.getByText(/r2, r3/)).not.toBeNull()
  })

  test("shows affected routeIds if routeIdAtCreation isn't relevant", () => {
    const n: Notification = {
      ...notificationWithMatchedVehicle,
      reason: "diverted",
      routeIds: ["route2", "route3"],
      routeIdAtCreation: "route1",
    }
    const result = render(
      <RoutesProvider routes={routes}>
        <NotificationCard
          notification={n}
          currentTime={new Date()}
          openVPPForCurrentVehicle={jest.fn()}
        />
      </RoutesProvider>
    )
    expect(result.getByText(/r2, r3/)).not.toBeNull()
  })

  const reasons: Record<NotificationReason, RegExp> = {
    manpower: /No Operator/,
    disabled: /Disabled/,
    diverted: /Diversion/,
    accident: /Accident/,
    other: /Other/,
    adjusted: /Adjusted/,
    operator_error: /Operator Error/,
    traffic: /Traffic/,
    chelsea_st_bridge_raised: /Chelsea St Bridge Raised/,
    chelsea_st_bridge_lowered: /Chelsea St Bridge Lowered/,
  }

  test.each(Object.keys(reasons) as NotificationReason[])(
    "renders notification with reason %s",
    (reason) => {
      const runIds =
        reason === "chelsea_st_bridge_raised" ||
        reason === "chelsea_st_bridge_lowered"
          ? []
          : notification.runIds

      const n: Notification = { ...notification, reason, runIds }
      const result = render(
        <RoutesProvider routes={routes}>
          <NotificationCard
            notification={n}
            currentTime={new Date()}
            openVPPForCurrentVehicle={jest.fn()}
          />
        </RoutesProvider>
      )

      expect(result.queryByText(reasons[reason])).not.toBeNull()
    }
  )

  test("clicking through opens VPP and hides notification", async () => {
    const updatedNotification = {
      ...notification,
      tripIds: ["123", "456", "789"],
    }
    const dispatch = jest.fn()
    const currentTime = new Date()
    const openVPPForCurrentVehicle = jest.fn()

    const user = userEvent.setup()
    const result = render(
      <RoutesProvider routes={routes}>
        <NotificationCard
          notification={updatedNotification}
          currentTime={currentTime}
          openVPPForCurrentVehicle={openVPPForCurrentVehicle}
          hideLatestNotification={() => dispatch(hideLatestNotification())}
          noFocusOrHover={true}
        />
      </RoutesProvider>
    )
    expect(openVPPForCurrentVehicle).not.toHaveBeenCalled()
    expect(dispatch).not.toHaveBeenCalled()

    await user.click(result.getByText(/run1/))

    expect(openVPPForCurrentVehicle).toHaveBeenCalled()
    expect(dispatch).toHaveBeenCalledWith({ type: "HIDE_LATEST_NOTIFICATION" })
  })

  test.each<{
    reason: NotificationReason
    should_fire_fs_event: boolean
  }>([
    { should_fire_fs_event: true, reason: "chelsea_st_bridge_raised" },
    { should_fire_fs_event: true, reason: "chelsea_st_bridge_lowered" },
    { should_fire_fs_event: false, reason: "other" },
    { should_fire_fs_event: false, reason: "manpower" },
    { should_fire_fs_event: false, reason: "disabled" },
    { should_fire_fs_event: false, reason: "diverted" },
    { should_fire_fs_event: false, reason: "accident" },
    { should_fire_fs_event: false, reason: "adjusted" },
    { should_fire_fs_event: false, reason: "operator_error" },
    { should_fire_fs_event: false, reason: "traffic" },
  ])(
    "clicking bridge notification should trigger FS event: %s",
    async ({ reason, should_fire_fs_event }) => {
      mockFullStoryEvent()
      const updatedNotification = notificationFactory.build({
        reason,
      })
      const dispatch = jest.fn()
      const currentTime = new Date()
      const openVPPForCurrentVehicle = jest.fn()

      const user = userEvent.setup()
      const result = render(
        <RoutesProvider routes={routes}>
          <NotificationCard
            notification={updatedNotification}
            currentTime={currentTime}
            openVPPForCurrentVehicle={openVPPForCurrentVehicle}
            hideLatestNotification={() => dispatch(hideLatestNotification())}
            noFocusOrHover={true}
          />
        </RoutesProvider>
      )
      expect(openVPPForCurrentVehicle).not.toHaveBeenCalled()
      expect(dispatch).not.toHaveBeenCalled()

      await user.click(result.getByText(/run1/))

      if (should_fire_fs_event) {
        expect(window.FS!.event).toHaveBeenCalledWith(
          "User clicked Chelsea Bridge Notification"
        )
      } else {
        expect(window.FS!.event).not.toHaveBeenCalledWith(
          "User clicked Chelsea Bridge Notification"
        )
      }
    }
  )
})
