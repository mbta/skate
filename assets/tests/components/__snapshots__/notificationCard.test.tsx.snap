// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationCard renders activated detour notification if user is in DetoursList group 1`] = `
<body>
  <div>
    <div
      aria-labelledby="card-label-:rh:"
      class="c-card c-card--kiwi"
    >
      <button
        class="c-card__left"
      >
        <div
          class="c-card__left-content"
        >
          <div
            class="c-card__top-row"
          >
            <div
              class="c-card__title"
              id="card-label-:rh:"
            >
              <span>
                <svg />
              </span>
              Detour - Active
            </div>
            <div
              class="c-card__time"
            >
              0 min
            </div>
          </div>
          <div
            class="c-card__contents"
          >
            <div
              class="c-card__body"
            >
              <div
                class="d-flex flex-row gap-2"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  4
                </div>
                <div>
                  <div
                    class="fw-semibold"
                  >
                    Headsign 4
                  </div>
                  <div
                    class="fw-normal text-body-secondary"
                  >
                    From 
                    Origin station 4
                  </div>
                  <div
                    class="fw-normal"
                  >
                    Outbound
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </button>
    </div>
  </div>
</body>
`;

exports[`NotificationCard renders detour deactivated notification if user is in DetoursList group 1`] = `
<body>
  <div>
    <div
      aria-labelledby="card-label-:ri:"
      class="c-card c-card--kiwi"
    >
      <button
        class="c-card__left"
      >
        <div
          class="c-card__left-content"
        >
          <div
            class="c-card__top-row"
          >
            <div
              class="c-card__title"
              id="card-label-:ri:"
            >
              <span>
                <svg />
              </span>
              Detour - Closed
            </div>
            <div
              class="c-card__time"
            >
              0 min
            </div>
          </div>
          <div
            class="c-card__contents"
          >
            <div
              class="c-card__body"
            >
              <div
                class="d-flex flex-row gap-2"
              >
                <div
                  class="c-route-pill c-route-pill--bus"
                >
                  1
                </div>
                <div>
                  <div
                    class="fw-semibold"
                  >
                    Headsign 1
                  </div>
                  <div
                    class="fw-normal text-body-secondary"
                  >
                    From 
                    Origin station 1
                  </div>
                  <div
                    class="fw-normal"
                  >
                    Outbound
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </button>
    </div>
  </div>
</body>
`;
